zrml/parimutuel/src/benchmarking.rs

```diff
@@ -0,0 +1,136 @@
+
+        let market_id = setup_market::<T>(MarketType::Categorical(64u16));
+
+        let amount = T::MinBetSize::get();
+        let asset = Asset::ParimutuelShare(market_id, 0u16);
+
+        let market = T::MarketCommons::market(&market_id).unwrap();
+        T::AssetManager::deposit(market.base_asset, &buyer, amount).unwrap();
+
+        #[extrinsic_call]
+        buy(RawOrigin::Signed(buyer), asset, amount);
+    }
+
+    #[benchmark]
+    fn claim_rewards() {
+        // max category index is worst case
+        let market_id = setup_market::<T>(MarketType::Categorical(64u16));
+
+        let winner = whitelisted_caller();
+        let winner_asset = Asset::ParimutuelShare(market_id, 0u16);
+        let winner_amount = T::MinBetSize::get() + T::MinBetSize::get();
+        buy_asset::<T>(market_id, winner_asset, &winner, winner_amount);
+
+        let loser = whitelisted_caller();
+        let loser_asset = Asset::ParimutuelShare(market_id, 1u16);
+        let loser_amount = T::MinBetSize::get();
+        buy_asset::<T>(market_id, loser_asset, &loser, loser_amount);
+
+        T::MarketCommons::mutate_market(&market_id, |market| {
+            market.status = MarketStatus::Resolved;
+            market.resolved_outcome = Some(OutcomeReport::Categorical(0u16));
+            Ok(())
+        })?;
+
+        #[extrinsic_call]
+        claim_rewards(RawOrigin::Signed(winner), market_id);
+    }
+
+    #[benchmark]
+    fn claim_refunds() {
+        // max category index is worst case
+        let market_id = setup_market::<T>(MarketType::Categorical(64u16));
+
+        let loser_0 = whitelisted_caller();
+        let loser_0_index = 0u16;
+        let loser_0_asset = Asset::ParimutuelShare(market_id, loser_0_index);
+        let loser_0_amount = T::MinBetSize::get() + T::MinBetSize::get();
+        buy_asset::<T>(market_id, loser_0_asset, &loser_0, loser_0_amount);
+
+        let loser_1 = whitelisted_caller();
+        let loser_1_index = 1u16;
+        let loser_1_asset = Asset::ParimutuelShare(market_id, loser_1_index);
+        let loser_1_amount = T::MinBetSize::get();
+        buy_asset::<T>(market_id, loser_1_asset, &loser_1, loser_1_amount);
+
+        T::MarketCommons::mutate_market(&market_id, |market| {
+            market.status = MarketStatus::Resolved;
+            let resolved_index = 9u16;
+            let resolved_outcome = OutcomeReport::Categorical(resolved_index);
+            assert_ne!(resolved_index, loser_0_index);
+            assert_ne!(resolved_index, loser_1_index);
+            let resolved_asset = Asset::ParimutuelShare(market_id, resolved_index);
+            let resolved_issuance_asset = T::AssetManager::total_issuance(resolved_asset);
+            assert!(resolved_issuance_asset.is_zero());
+            market.resolved_outcome = Some(resolved_outcome);
+            Ok(())
+        })?;
+
+        #[extrinsic_call]
+        claim_refunds(RawOrigin::Signed(loser_0), loser_0_asset);
+    }
```

Comment:
Missing `verify` statement that checks that `loser_0` and `loser_1` received the refund (call successful).

Reply:
I personally am in favour of doing this in unit tests. This assertion is generally not really required in the benchmarks. Benchmarks are required to succeed in the call.

Reply:
I personally am in favour of doing this in unit tests. This assertion is generally not really required in the benchmarks. Benchmarks are required to succeed in the call.

Reply:
To assert in benchmarks is especially useful if we want to ensure a specific benchmark path.